### ISSUE TITLE
retinet (or focal loss) accuracy issue

### ISSUE THREAD

**Issue by jwnsu** [facebookresearch/Detectron#93]
Anyone able to reproduce retinet (FPN) accuracy, on any dataset?

Tried it on a new dataset and got VOC2012 accuracy around 0.52 mAP (ResNet50). Same dataset with mask faster rcnn FPN (ResNet50) from this repository got >0.88 mAP (dataset settings are identical in both, SCALES/NMS too). Learn rates adjustment does not help. Plan to change more parameters (alpha, gamma etc.)

Further inspection shows 6 to 7 objects got 0 mAP in retinet case, very unusual. 

Debug suggestions are welcome.

**Comment by rbgirshick** [comment]
The fact that some classes have 0 AP is very unusually indeed and makes it sound like there's a bug somewhere (we have actually never tried training and testing on VOC). Random guess: I would look into places that depend on the number of classes in the dataset (e.g., where class score are computed for each anchor). Perhaps there is something that is not handled properly when going from the 81 classes in COCO to the 21 in VOC?

**Comment by jwnsu** [comment]
lib/roi_data/retinanet seems to get # of classes from dataset input.

coco dataset stuck at low 0.11 mmAP too. Likely some bugs in data preparation logic.

Is retinanet code path in this repository tried on a dataset (for accuracy)? The code base here seems different from Focal Loss paper (arxiv 1708.02002), e.g. FPN cannot be turned off in this repository.

**Comment by tylin** [comment]
From the description the mAP of non-zero categories is 0.8 given 7 categories have 0 AP. It looks like there are some mapping error of category ids for those 7 categories.

**Comment by jwnsu** [comment]
Thx for the info. We tried the same coco instances files to train with mask rcnn fpn, accuracy from mask fpn is very good. It seems retinet code path uses its own data preparation (roi_data/retinanet), some bugs there?

**Comment by wgting96** [comment]
I am new to Caffe2 framework. I am not sure if the retinanet misses to set the number of classes?

https://github.com/facebookresearch/Detectron/blob/021685d42f7e8ac097e2bcf79fecb645f211378e/lib/modeling/retinanet_heads.py#L280-L304 

Should it be something like the following codes if I would like to train the retinanet on my custom dataset?

```python
            cls_focal_loss, gated_prob = model.net.SoftmaxFocalLoss(
                [
                    cls_lvl_logits, 'retnet_cls_labels_' + suffix,
                    'retnet_fg_num'
                ],
                ['fl_{}'.format(suffix), 'retnet_prob_{}'.format(suffix)],
                gamma=cfg.RETINANET.LOSS_GAMMA,
                alpha=cfg.RETINANET.LOSS_ALPHA,
                num_classes=model.num_classes,
                scale=model.GetLossScale(),
            )
```

**Comment by kampelmuehler** [comment]
Any news on this? Struggling with the same behavior on ILSVRC

**Comment by jwnsu** [comment]
@wgting96 your suggestion does not seem to make a difference. After 11 iterations, 7 (out of about 40) classes have 0 mAP for retinanet (ResNet50), including first 6 classes are all 0 mAP. In comparison, mask knn (same ResNet50) needs only 2000 iterations to get good mAP across all classes, with >0.7 overall mAP (each of first 6 classes are all above 0.7 mAP).

**Comment by kampelmuehler** [comment]
@jwnsu unfortunately you are correct. No luck here after 20k iterations either.

**Comment by twmht** [comment]
@rbgirshick 

there is no test for `sigmoidFocalLoss layer`, I tried to migrate this implementation to caffe, and write a gradient checker for that, but got some of big gradient difference ( the gradient difference between analytical gradient and numerical gradient are greater than 0.01) .

So I doubt there is some bugs in the `sigmoidFocalLoss layer`.

**Comment by jwnsu** [comment]
The accuracy is much improved (from 0.5 to >0.75), however, there are still 4 objects with 0 mAP (reduced from 7 to 8 0 mAP previously), those 4 objects all have >0.9 mAP with faster rcnn. Tried ResNxt 32x8, the same 4 objects still have 0 mAP. Likely, either retinanet's training data generation or focus training does not generate training for those 4 objects.  The 4 0-mAP objects have id 3,4,5,8.
